omnisharp-dotnet/src/SonarLint.OmniSharp.Plugin/DiagnosticWorker/SonarAnalyzerAssembliesProvider.cs

```
@@ -30,4 +34,57 @@ internal interface ISonarAnalyzerAssembliesProvider
     {
         ImmutableArray<Assembly> Assemblies { get; }
     }
+
+    [Export(typeof(ISonarAnalyzerAssembliesProvider))]
+    [PartCreationPolicy(CreationPolicy.Shared)]
+    internal class SonarAnalyzerAssembliesProvider : ISonarAnalyzerAssembliesProvider
+    {
+        /// <summary>
+        /// It is the responsibility of the java plugin to place the analyzer assemblies in the "analyzers" sub directory.
+        /// </summary>
+        internal static string AnalyzersDirectory { get; } =
+            Path.Combine(Path.GetDirectoryName(typeof(SonarAnalyzerAssembliesProvider).Assembly.Location), "analyzers");
+        
+        private readonly IAssemblyLoader loader;
+        private readonly Func<string, string[]> getFilesInDirectory;
+        private ImmutableArray<Assembly> loadedAssemblies;
+
+        public ImmutableArray<Assembly> Assemblies
+        {
+            get
+            {
+                if (loadedAssemblies == null)
+                {
+                    loadedAssemblies = LoadAssemblies();
+                }
+
+                return loadedAssemblies;
+            }
+        }
+
+        [ImportingConstructor]
+        public SonarAnalyzerAssembliesProvider(IAssemblyLoader loader)
+            : this(loader, Directory.GetFiles)
+        {
+        }
+
+        internal SonarAnalyzerAssembliesProvider(IAssemblyLoader loader, Func<string, string[]> getFilesInDirectory)
+        {
+            this.loader = loader;
+            this.getFilesInDirectory = getFilesInDirectory;
+        }
+
+        private ImmutableArray<Assembly> LoadAssemblies()
+        {
+            var builder = ImmutableArray.CreateBuilder<Assembly>();
+
+            foreach (var filePath in getFilesInDirectory(AnalyzersDirectory))
+            {
+                var analyzerAssembly = loader.LoadFrom(filePath);
```

I'm not sure what to do with exception handling.
Omnisharp's assembly loader will catch all exceptions and just return a null.
https://github.com/OmniSharp/omnisharp-roslyn/blob/master/src/OmniSharp.Host/Services/AssemblyLoader.cs#L85

I'm not sure what our behavior should be: throw an exception and let it propagate to our worker and our endpoint?
Perhaps we should revisit it later when we have all the pieces in place.

---

> throw an exception and let it propagate to our worker and our endpoint?

We want the failure to be obvious to the Java code during development.

I don't think we need to do anything special if the directory doesn't exist if that already throws an exception that will be propagated through the service.

I think we should throw if we don't find any assemblies; if we don't, it will be harder to track down from the Java.

---

Also: let's throw on null assemblies too. We don't think it should happen if the plugin packaging is correct. However, there might be cross-platform cases where that we haven't though of/tested that will cause failures in practice. In that case, failing fast would help us and users.